### PR TITLE
Adds support for SageMaker Processing Jobs

### PR DESCRIPTION
**Pull request by edgar-t**
# TL;DR
- Simplest method to add support for creating and using sagemaker processing jobs in a Flyte workflow 

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [x ] Plugin

## Open Questions
- Currently built and pushed the python plugin wheel to CodeArtifact manually
  -  should this be completed via a Jenkinsfile

---
❗ ❗ This does not follow the current implementation of the aws-sagemaker plugin as this is a flytekit-only plugin versus a native backend plugin❗ ❗ 

[flytekit-only-task-plugin](https://docs.flyte.org/projects/cookbook/en/latest/auto/core/extend_flyte/index.html#flytekit-only-task-plugin)

## What is the recommended approach suggested by Flyte
In short this plugin would be considered a light-weight extension versus the current performant multi-language sagemaker extension 
image

Since it is a light-weight extension, the source code required to enable the processing job can be summed up to the
-  `plugins/flytekit-aws-sagemakerprocessing` directory

If we desire to modify this plugin to conform with the sagemaker plugin here are the changes required (**afaict):
- Create a new `processing.proto` file in https://github.com/flyteorg/flyteidl/tree/master/protos/flyteidl/plugins/sagemaker
- Use [flyteidl repo's](https://github.com/flyteorg/flyteidl) `Makefile` to use the protocol buffer compiler to autogenerate the `.py`, similar to https://github.com/flyteorg/flyteidl/blob/master/gen/pb_python/flyteidl/plugins/sagemaker/training_job_pb2.py
- Merge these changes into [flyteidl repo](https://github.com/flyteorg/flyteidl)
- Create a new `python SageMaker Model` file in [flytekit repo](https://github.com/flyteorg/flytekit) similar to [training_job.py](https://github.com/flyteorg/flytekit/blob/master/plugins/flytekit-aws-sagemaker/flytekitplugins/awssagemaker/models/training_job.py)
- Create a new `python Processing` file in [flytekit repo](https://github.com/flyteorg/flytekit) similar to [training.py](https://github.com/flyteorg/flytekit/blob/master/plugins/flytekit-aws-sagemaker/flytekitplugins/awssagemaker/training.py)
- Merge [flytekit repo](https://github.com/flyteorg/flytekit) changes
- Create new GO `Processing` API specification using the sagemaker-operator in [flyteplugins](https://github.com/flyteorg/flyteplugins), similar to [training.go](https://github.com/flyteorg/flyteplugins/blob/master/go/tasks/plugins/k8s/sagemaker/builtin_training.go)
  - 🔔 This is as far as i got down this initial rabbit hole 🔔  
- Merge [flyteplugins](https://github.com/flyteorg/flyteplugins) changes
- Update the default plugins loaded by [flytepropeller repo](https://github.com/flyteorg/flytepropeller/blob/master/plugins/loader.go)
- Merge flytepropeller changes
- Build and store new flytepropeller image in ECR
- Update Flyte helm chart to use this updated image
- Enable new plugin via flyte deployment overrides in [values-eks.yaml](https://github.com/flyteorg/flyte/blob/master/charts/flyte-core/values-eks.yaml#L268) 
- Redeploy Flyte

### CODE REVIEW

File: plugins/flytekit-aws-sagemakerprocessing/flytekitplugins/sagemakerprocessing/processing.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+            entrypoint=None,
+            env=self.task_config.app_specification.environment
+        )
+        job_inputs = [ProcessingInput(input_name=x.name, destination=x.local_path, source=x.s3_uri) for x in
+                      kwargs.get("inputs")] if kwargs.get("inputs") and type(kwargs.get("inputs")) is not str else []
+        job_outputs = [ProcessingOutput(output_name=x.name, destination=x.s3_uri, source=x.local_path) for x in
+                       kwargs.get("outputs")] if kwargs.get("outputs") and type(kwargs.get("outputs")) is not str else []
+
+        processor.run(
+            job_name=self.name + datetime.now().strftime("%m-%d-%H-%M-%S"),
```
Comment (lou-k) [inline plugins/flytekit-aws-sagemakerprocessing/flytekitplugins/sagemakerprocessing/processing.py:132]: optional: add a `-` between `name` and the datestring

File: plugins/flytekit-aws-sagemakerprocessing/flytekitplugins/sagemakerprocessing/processing.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            for output in kwargs.get("outputs"):
+                files.append(FlyteFile(output.s3_uri))
+            return files
+        elif self.output_type is OutputType.DIRECTORY:
+            path, ext = os.path.splitext(kwargs.get("outputs")[0].s3_uri)
+            if ext:
+                path = os.path.dirname(path)
+            return FlyteDirectory(path)
+        else:
+            return "Processing Job completed with no return value"
```
Comment (lou-k) [inline plugins/flytekit-aws-sagemakerprocessing/flytekitplugins/sagemakerprocessing/processing.py:151]: should this return `None` instead of a `str`?

**Review by lou-k — APPROVED** [review]
nice work here!

Do we have cicd for this repo? Would be nice to have this published to code artifact automatically (I saw you pushed it already for the examples)

File: plugins/flytekit-aws-sagemakerprocessing/flytekitplugins/sagemakerprocessing/processing.py (hunk 2 above)
Comment (edgar-t) [reply plugins/flytekit-aws-sagemakerprocessing/flytekitplugins/sagemakerprocessing/processing.py:161]: No, since the interface still needs to define a output type besides `None`, and I set it as a `str` [for those cases](https://github.com/curalate/flytekit/blob/fa2431e6fe7601aa2a262cb02b7a1aeb8feda9f0/plugins/flytekit-aws-sagemakerprocessing/flytekitplugins/sagemakerprocessing/processing.py#L100), so rather than a `None` any string will work

**Comment by edgar-t** [comment]
@lou-k Added CICD support for the wheel upload to codeartifact
